### ISSUE TITLE
Webpack 4 compatibility

### ISSUE THREAD

**Issue by errorx666** [webpack-contrib/extract-text-webpack-plugin#701]
I'm trying to use this plugin with webpack 4 alpha 5 and getting the following error:

```
Error: Chunk.entrypoints: Use Chunks.groupsIterable and filter by instanceof Entrypoint instead
    at Chunk.get (C:\Users\errorx666\Documents\src\node_modules\webpack\lib\Chunk.js:463:9)
    at C:\Users\errorx666\Documents\src\node_modules\extract-text-webpack-plugin\dist\index.js:176:48
    at Array.forEach (<anonymous>)
    at C:\Users\errorx666\Documents\src\node_modules\extract-text-webpack-plugin\dist\index.js:171:18
    at AsyncSeriesHook.eval [as callAsync] (eval at create (C:\Users\errorx666\Documents\src\node_modules\webpack\node_modules\tapable\lib\HookCodeFactory.js:24:12), <anonymous>:7:1)
    at AsyncSeriesHook.lazyCompileHook [as _callAsync] (C:\Users\errorx666\Documents\src\node_modules\webpack\node_modules\tapable\lib\Hook.js:35:21)
    at Compilation.seal (C:\Users\errorx666\Documents\src\node_modules\webpack\lib\Compilation.js:787:27)
    at hooks.make.callAsync.err (C:\Users\errorx666\Documents\src\node_modules\webpack\lib\Compiler.js:460:17)
    at _err0 (eval at create (C:\Users\errorx666\Documents\src\node_modules\webpack\node_modules\tapable\lib\HookCodeFactory.js:24:12), <anonymous>:11:1)
    at _addModuleChain (C:\Users\errorx666\Documents\src\node_modules\webpack\lib\Compilation.js:672:11)
    at processModuleDependencies.err (C:\Users\errorx666\Documents\src\node_modules\webpack\lib\Compilation.js:614:8)
    at process._tickCallback (internal/process/next_tick.js:150:11)
```

**Comment by michael-ciniawsky** [comment]
Yep `extract-text-webpack-plugin` isn't updated for `webpack >= v4.0.0` yet and is completely broken since `webpack@4.0.0-alpha.5`. It may be superseded by another plugin/solution or will get an update before the final `webpack` v4.0.0 release

**Comment by apfelbox** [comment]
Is there any update and/or planning going on?
I am still wondering what the recommended way for compiling (S)CSS is. I previously used this plugin and would like to switch to webpack 4, but without compilation of CSS this isn't possible.

**Comment by karmajunkie** [comment]
Yes, could we have an update on this? I'm excited about the v4 improvements, but honestly, this is a pretty fundamental plugin for a lot of people, I don't see how you go into beta before this one is ready.  Please advise.

**Comment by montogeek** [comment]
@karmajunkie This is not webpack library, it is supported by contributors. If you want it to be ready for webpack 4 beta, please contribute to it.

**Comment by errorx666** [comment]
@karmajunkie If it's urgent that this issue be resolved, and you don't think you can resolve it yourself, you could post a bounty for it:
https://www.bountysource.com/issues/54124099-webpack-4-compatibility

**Comment by evilebottnawi** [comment]
@errorx666 please use opencollective for this (https://opencollective.com/webpack)

**Comment by karmajunkie** [comment]
@montogeek that's fair, i was under the impression that being under webpack-contrib meant that it was an officially supported project by the webpack team. Given its important position with the webpack ecosystem it really should be. It boggles mind that they would put version 4 into beta without these kinds of "tier 1" plugins being updated, much less announce a 1 month beta period before labeling it production ready.

 I would love to help but I'm not familiar enough with webpack internals or plugin  API to do so. From the comment above it sounds like there is more clarity needed from either this project's maintainers or that of the webpack team on how best to move forward which makes it doubly difficult for new contributors to jump into an update.

**Comment by montogeek** [comment]
I am working on it, @sokra changed how it works internally (https://github.com/webpack/webpack/commit/18ae73dad4bb761593a9f5529a182601e1b57520#diff-1064caa9219b7da75fbad061b1a25ec5), breaking/big change. This will take time

**Comment by GuillaumeCisco** [comment]
Hey @montogeek thanks for taking a look on this issue. I'd like to warn you the diff you mentioned is not similar to the last version of Chunk.js : https://github.com/webpack/webpack/blob/next/lib/Chunk.js
Hope it will help to solve this issue ;)

**Comment by izaakschroeder** [comment]
@montogeek no promises, but you can have a look at https://github.com/faceyspacey/extract-css-chunks-webpack-plugin/pull/61/commits/4b3b0a55e2cf70bd9a4f756eaa3e27f6e0021509. It seems to work for me.

**Comment by montogeek** [comment]
@michael-ciniawsky told me that there is another plugin that is going to replace this one eventually, also this is could be support natively by webpack (https://github.com/webpack/webpack/pull/6448)

@izaakschroeder Does it pass the tests? This week prove to be very busy, couldn't continue working on it.

**Comment by izaakschroeder** [comment]
@montogeek I haven't tested it on this plugin, only the other one. Native CSS modules seem great though didn't realize they were landing in 4.x. 🚀

**Comment by qkdreyer** [comment]
It's out now: https://github.com/webpack/webpack/releases/tag/v4.0.0

**Comment by cklmercer** [comment]
@koddr looks like you moved the plugin but didn't remove loaders that reference the plugin.

**Comment by karolyi** [comment]
@cklmercer same here. a lot of other plugins are using this plugin.

also, I've spent 2 hrs to figure out how I can update my now completely broken frontend tooling, before I gave up and decided to rollback and lay low until the other extra modules (I'm looking at you, `stylus-loader` and `postcss-loader`) are working.

so now, at least I can work.

**Comment by cklmercer** [comment]
@karolyi yup.. same here. We're going to be waiting things out. Still super excited though. Love the "sensible defaults". I've actually been working on a webpack overhaul for my company for the past week or so that's about to go live next weeks, so I figured I'd at least test out webpack4 beforehand. Definitely going to have to wait and double back though. Not even mad..

**Comment by adityapurwa** [comment]
So, did webpack now have built in of something like `extract-text-webpack-plugin`, or its just that the plugin doesn't work with webpack@4?

**Comment by cklmercer** [comment]
@adityapurwa afaik, we're just waiting on the plugin to be updated to support v4.

**Comment by adityapurwa** [comment]
CMIIW

> Historically, JavaScript has been the only first-class module type in webpack. This caused a lot of awkward pains for users where they would not be able to effectively have CSS/HTML Bundles, etc. We have completely abstracted the JavaScript specificity from our code base to allow for this new API.

Taken from https://medium.com/webpack/webpack-4-released-today-6cdb994702d4

Does this mean now that webpack can accept CSS file and output it?

**Comment by bmayen** [comment]
@adityapurwa, From the "What's Next" (v4.x, v5) section of that article:
"CSS Module Type — CSS as Entry (Goodbye ExtractTextWebpackPlugin)"

**Comment by manix** [comment]
There is an [alpha release](https://github.com/webpack-contrib/extract-text-webpack-plugin/releases/tag/v4.0.0-alpha.0) of the plugin, for anyone as oblivious as me.

**Comment by brigand** [comment]
@manix the alpha version seems to be working for me on a fairly large project that relies on CSS order heavily.

**Comment by francoisromain** [comment]
@manix How do you install the alpha?

**Comment by jackwilsdon** [comment]
@francoisromain It's in the description of the release;
>     npm i -D extract-text-webpack-plugin@next

**Comment by MadOgre** [comment]
@evilebottnawi, errorx666 asked a legitimate question. There is no need to be passive-aggressive and demand money. Especially since the product doesn't work!

**Comment by evilebottnawi** [comment]
@MadOgre sorry, I made a mistake (the wrong tab was opened and it is reply for other issue) when I wrote this and noticed only now, I apologize

**Comment by evilebottnawi** [comment]
Also please look on @jackwilsdon answer above

**Comment by MadOgre** [comment]
@evilbottnawi :) simple misunderstanding. No harm done. Thanks for clearing it up.

**Comment by btakita** [comment]
Still an issue with Webpack 4.0.1...

**Comment by robcresswell** [comment]
@btakita Did you try `extract-text-webpack-plugin@next`?

**Comment by gempir** [comment]
If you plan to run your code in production I wouldn't recommend webpack 4 at this time. A lot important plugins are still in testing/upgrading phase. 

If you must use the Alpha like rob mentioned.

**Comment by ngohungphuc** [comment]
Try with `extract-text-webpack-plugin@next` and the issue happend  [here](https://github.com/webpack-contrib/extract-text-webpack-plugin/issues/731)

**Comment by btakita** [comment]
I'm getting the following error with `extract-text-webpack-plugin@next` (`4.0.0-beta.0`).

I don't want to complain, but I'm really confused over this upgrade webpack to version 4. How do we know when it's going to be ok to upgrade? The last thing I want to see is some random issue happen in production that is not caught in our `staging` or `uat` environments. There's a bunch of plugins being used by webpack, so it's difficult to track down all of the issues.

```
$ bin/webpack-dev-server 
 10% building modules 2/2 modules 0 active(node:1525) DeprecationWarning: Tapable.apply is deprecated. Call apply on the plugin directly instead
(node:1525) DeprecationWarning: Tapable.plugin is deprecated. Use new API on `.hooks` instead
ℹ ｢wds｣: Project is running at http://localhost:3035/
ℹ ｢wds｣: webpack output is served from /assets/packs/
ℹ ｢wds｣: Content not from webpack is served from /home/user/project/public/assets/packs
ℹ ｢wds｣: 404s will fallback to /index.html
 11% building modules 14/19 modules 5 active ...node_modules/querystring-es3/index.js(3:4) – A11y: <a> element should have an href attribute
 11% building modules 15/19 modules 4 active ...node_modules/querystring-es3/index.js(4:4) – A11y: <a> element should have an href attribute
 12% building modules 20/30 modules 10 active ...ode_modules/webpack/buildin/global.js(44:4) – A11y: <a> element should have an href attribute
 77% module and chunk tree optimization unnamed compat plugin/home/user/project/node_modules/webpack/lib/Chunk.js:460
		throw new Error(
		^

Error: Chunk.entrypoints: Use Chunks.groupsIterable and filter by instanceof Entrypoint instead
    at Chunk.get (/home/user/project/node_modules/webpack/lib/Chunk.js:460:9)
    at /home/user/project/node_modules/@rails/webpacker/node_modules/extract-text-webpack-plugin/dist/index.js:176:48
    at Array.forEach (<anonymous>)
    at /home/user/project/node_modules/@rails/webpacker/node_modules/extract-text-webpack-plugin/dist/index.js:171:18
    at AsyncSeriesHook.eval [as callAsync] (eval at create (/home/user/project/node_modules/tapable/lib/HookCodeFactory.js:24:12), <anonymous>:14:1)
    at AsyncSeriesHook.lazyCompileHook [as _callAsync] (/home/user/project/node_modules/tapable/lib/Hook.js:35:21)
    at Compilation.seal (/home/user/project/node_modules/webpack/lib/Compilation.js:881:27)
    at hooks.make.callAsync.err (/home/user/project/node_modules/webpack/lib/Compiler.js:464:17)
    at _done (eval at create (/home/user/project/node_modules/tapable/lib/HookCodeFactory.js:24:12), <anonymous>:9:1)
    at _err0 (eval at create (/home/user/project/node_modules/tapable/lib/HookCodeFactory.js:24:12), <anonymous>:20:22)
    at _addModuleChain (/home/user/project/node_modules/webpack/lib/Compilation.js:749:12)
    at processModuleDependencies.err (/home/user/project/node_modules/webpack/lib/Compilation.js:688:9)
    at process._tickCallback (internal/process/next_tick.js:112:11)
```

**Comment by btakita** [comment]
It seems like `@rails/webpacker` is locked on a certain version of `extract-text-webpack-plugin`. To fix, I added to 'package.json':

```
  "resolutions": {
    "**/**/extract-text-webpack-plugin": "^4.0.0-beta.0"
  }
```

That being said, I'm going to take the advice from @gempir & downgrade to v3.

**Comment by strflw** [comment]
> How do we know when it's going to be ok to upgrade?

Maybe whent the version is not a "alpha" or "beta" anymore…but thats just an uneducated guess…

**Comment by amrshakya** [comment]
can anyone provide me the documentation of webpack 3? coz in webpack website i just found the documentation for webpack 4

**Comment by amrshakya** [comment]
actually i wanted to use webpack 3 but couldn't find the doc for webpack 3

**Comment by amrshakya** [comment]
extract-text-webpack-plugin is available for webpack 1 2 and 3 right?

**Comment by kvasa** [comment]
You can try to use official plugin for webpack 4 directly from sokra -> https://github.com/webpack-contrib/mini-css-extract-plugin I guess it will replace this plugin in the future it's still alpha and there are some issues but it works.

**Comment by amrshakya** [comment]
thank you kvasa

**Comment by amrshakya** [comment]
but i wanted to use webpack 3 still it will be more stable since i need to use webpack in production but i didnt found the doc for webpack 3

**Comment by montogeek** [comment]
This is the page for webpack 3 https://webpack.js.org/plugins/extract-text-webpack-plugin/

**Comment by Vinnl** [comment]
@amrshakya Please ask questions on StackOverflow or the Webpack Gitter: https://gitter.im/webpack/webpack

There are quite a few people subscribed to this issue so they get notified when this plugin has a new Webpack 4-compatible release, and now they're being drowned in notification emails of your comments.

(And: if people on Gitter/StackOverflow tell you there is no way to access old Webpack docs, you might want to [report an issue against Webpack](https://github.com/webpack/webpack/issues/new).

**Comment by brigand** [comment]
From what I can tell, I was getting some missing rules in the resulting CSS. Disabled extract text and it worked fine. Tried with the latest `@next` on webpack 4.0.0 and 4.1.0.

**Comment by errorx666** [comment]
I'm switching to [`mini-css-extract-plugin`](/webpack-contrib/mini-css-extract-plugin).
I'll no longer be monitoring this issue.  The package maintainer can close it when/if it's resolved.

**Comment by bstaruk** [comment]
`4.0.0-beta.0` resolves this issue for me-- testable on [starbase](https://github.com/bstaruk/starbase) (dev branch for now).

**Comment by JohnGalt1717** [comment]
doesn't resolve the issue for me. the resulting css files are 0 length.

**Comment by snaerth** [comment]
What's the status on this bug?

**Comment by evilebottnawi** [comment]
Just for information, in near future `webpack` implements buind-in support `css`, right now we have https://github.com/webpack-contrib/mini-css-extract-plugin POC. We will fix bugs here, but better to migrate on `mini-css-extract-plugin`, same logic was inside `webpack@4`. More users == more issue == more fixes == better support `css` in webpack :+1:

**Comment by pcostarg** [comment]
@evilebottnawi how can we use mini-css-extract-plugin with SCSS? I was trying something like this:

`{
                // Rules for SCSS files. Loaders run in reverse order
                test: /\.(scss)$/,
                    use: [{
                        loader: 'MiniCssExtractPlugin.loader'
                    },
                        {
                        loader: "css-loader"
                    }, {
                        loader: "sass-loader",
                        options: {
                            sourceMap: true,
                            precision: 8,
                            data: "$ENV: " + "DEVELOP" + ";"
                        },
                    }],
            },`

**Comment by pcostarg** [comment]
Nevermind, found out what was wrong with it. It's as simple as it is below and seems to work fine, dumping extract-text-webpack-plugin for mini-css-extract-plugin

Remember to follow everything else from the readme on the project and setup the rules like this (you can ignore the sass-loader options if they are not needed for you):
```
rules: [
            {
                // Rules for SCSS files. Loaders run in reverse order
                test: /\.(scss)$/,
                use: [
                    MiniCssExtractPlugin.loader,

                    {
                        loader: "css-loader"
                    }, {
                        loader: "sass-loader",
                        options: {
                            sourceMap: true,
                            precision: 8,
                            data: "$ENV: " + "PRODUCTION" + ";"
                        },
                    }],
            },
        ],
```

**Comment by sathify** [comment]
@pcostarg what does this `data: "$ENV: " + "PRODUCTION" + ";"` mean?

**Comment by brigand** [comment]
@sathify it's injecting a variable into the scss code, like DefinePlugin does for JS in webpack.

**Comment by pcostarg** [comment]
@sathify with the data option you can inject data into your SASS files using the sass-loader. That was still static hard coded. Now I have `data: "$ENV: " + $env.name + ";"` so that I can use the webpack-cli to send the environment that I want to build.

This was necessary in my project because we have many different environments and they have some small differences in the CSS. By sending the environment to SASS I can control those small differences with ease.

For more information check the sass-loader documentation near the end about Environment variables: https://github.com/webpack-contrib/sass-loader

Also the webpack documentation about environment variables: https://webpack.js.org/guides/environment-variables/

**Comment by michael-ciniawsky** [comment]
TL;DR Use [`mini-css-extract-plugin`](https://github.com/webpack-contrib/mini-css-extract-plugin) instead :)

The `webpack >= v4.0.0` 'support' for `extract-text-webpack-plugin` was moved to [`mini-css-extract-plugin`](https://github.com/webpack-contrib/mini-css-extract-plugin) as more lightweight approach for `webpack >= v4.0.0` and ETWP entered maintenance mode. We are sorry for the inconvenience here, but ETWP reached a point where maintaining it become too much of a burden and it's not the first time upgrading a major `webpack` version was complicated and cumbersome due to issues with ETWP. The mid-term plan is integrate CSS support directly into `webpack` to avoid this kind of issue in the future

**Comment by eliduke** [comment]
I'm not totally sure how to "use" `mini-css-extract-plugin`. Pretty new to webpack, so let me break it down here.

This is my `package.json` file:
```
{
  "dependencies": {
    "@rails/webpacker": "^3.3.1",
    "coffeescript": "1.12.7",
    "lodash": "^4.17.4",
    "vue": "^2.5.13",
    "vue-loader": "^14.1.1",
    "vue-template-compiler": "^2.5.13",
    "vuex": "^2.4.0"
  },
  "devDependencies": {
    "uglify-js": "https://github.com/mishoo/UglifyJS2#harmony-v2.8.22",
    "webpack": "^4.1.0",
    "webpack-cli": "^2.0.12",
    "webpack-dev-server": "^3.1.1"
  }
}
```
And I'm using  `yarn install`.

I'm running a rails app in docker using `supervisord` to spin up the rails server and the webpack-dev-server. When I fire everything up, I'm getting the classic error from up above:
```
Error: Chunk.entrypoints: Use Chunks.groupsIterable and filter by instanceof Entrypoint instead
```
But, now I don't know what to do. It sounds like I'm supposed to use `mini-css-extract-plugin`, but where does that go? I've tried adding it to the dependencies _and_ devDependencies but I'm not having any luck. Each time I wipe out the `node_modules` folder and `yarn.lock` and then `yarn install` again. When I fire everything up, same error.

Thoughts? Suggestions?

**Comment by leebenson** [comment]
@eliduke - you need to put it in your webpack config -> https://github.com/webpack-contrib/mini-css-extract-plugin#configuration

If you don't know what that is: https://webpack.js.org/concepts/configuration/

**Comment by eliduke** [comment]
@leebenson Thanks for the quick response, but I'm still a little lost. This is an existing code-base with a new webpack implementation for getting some Vue components up and running. I'm trying to dig in and figure stuff out, but I actually don't have a `webpack.config.js` as far as I can tell. This is a rails app and I'm using the `webpacker` gem. During the install rake task it creates the following files:
```
.postcssrc.yml
.babelrc
config/webpacker.yml
config/webpack/environment.js
config/webpack/development.js
config/webpack/production.js
config/webpack/test.js
app/javascript/packs/application.js
```
Am I supposed to create the `webpack.config.js` manually?

**Comment by edmorley** [comment]
@eliduke it sounds like you are using webpacker, so you'll need to follow up with that project for webpack 4 compatibility. See: rails/webpacker/issues/1224

**Comment by kumaresan-subramani** [comment]
Error: Chunk.entrypoints: Use Chunks.groupsIterable and filter by instanceof Entrypoint instead

this is my error and i'm little new to web-pack ,i need suggestions to rectify it

**Comment by andrewtanner1987** [comment]
@kumaresan-subramani See here https://github.com/webpack-contrib/extract-text-webpack-plugin/issues/760

I used `npm i -D extract-text-webpack-plugin@next` and it works fine for me. Note that this is the beta release though.

**Comment by SrShark** [comment]
execute `npm install -D extract-text-webpack-plugin@next` that solve the problema. I'm using `"extract-text-webpack-plugin": "^4.0.0-beta.0"`

**Comment by kumaresan-subramani** [comment]
@andrewtanner1987  thanks, it works.....
 i need one more help, can you please give step by step procedure for to update web-pack version 3.6.0 to 4.0.0 and hoe to clear errors which are occurring during updation

**Comment by andre-bahia** [comment]
if none of these solutions work, you can try the commnads below:

Before executing the commands, be sure to have installed webpack 4 into `devDependencies` of your `package.json`
```json
"devDependencies": {
    "webpack": "^4.4.0",
    "webpack-cli": "^2.0.13"
}
```

```
npm cache clean --force

yarn cache clean

rm yarn.lock

rm package-lock.json

rm -rf /node_modules

yarn

npm i -D extract-text-webpack-plugin@next
```

**Comment by alexratiu** [comment]
Ty @jackwilsdon that's work for me `npm i -D extract-text-webpack-plugin@next`

**Comment by pleerock** [comment]
with `@next` it does not work for me in lazily loaded chunks

**Comment by pod4g** [comment]
mark

**Comment by pod4g** [comment]
run `npm i -D extract-text-webpack-plugin@next` worked, thanks

**Comment by kellyrmilligan** [comment]
that got rid of my error, but the build just stalls.   is there any change to how it is configured for the `@next` branch?

**Comment by jonny64bit** [comment]
I fixed my issues with this by moving over to the [mini-css-extract-plugin](https://github.com/webpack-contrib/mini-css-extract-plugin)

**Comment by leebenson** [comment]
In case it's helpful to anyone, I've just finished creating a [CSS preset for Launch.js](https://github.com/launchjs/styles) that uses mini-css-extract-plugin.

While it's specific to the Launch.js framework, here's the [pertinent config](https://github.com/launchjs/styles/blob/master/src/styles.ts) that should hopefully help some devs navigate using the new plugin in development (with hot reloading), as well as how to create a single chunk in a comparable fashion to how extract-text-webpack-plugin does it.

I spent a bit of time figuring out edge cases over the weekend, and there's loaders there for PostCSS, Sass and Less too, in case that's of any help.

**Comment by flo-pereira** [comment]
`yarn add --dev extract-text-webpack-plugin@next`

**Comment by MartinX3** [comment]
@flo-pereira 
That was already mentioned several times.

**Comment by josephecombs** [comment]
Those of you using the latest of Rails' webpacker, make sure your package.lock has the following line (or equivalent):

`"@rails/webpacker": "^4.0.0.pre.pre.2",`

I tried the solutions in this issue, but it turned out the old version of webpacker depended on the messed up version of extract-text-webpack-plugin

**Comment by bradleyayers** [comment]
@josephecombs do you mean `^4.0.0-pre.2`?

**Comment by josephecombs** [comment]
I do not - that is is the exact name of the rails gem, and corresponds to how package.json looks after you run `rails webpacker:install`. It also will ultimately resolve to `^4.0.0-pre.2` in yarn.lock so this is a moot point.

**Comment by bradleyayers** [comment]
Gotcha! Interestingly when I ran `yarn add @rails/webpacker@^4.0.0.pre.pre.2` it prompted me to select a version that exists (apparently `4.0.0.pre.pre.2`) did not exist, and the first option was `4.0.0-pre.2`.

**Comment by josephecombs** [comment]
yes ^^ I went through this again and received the same prompt.

**Comment by Eddybag** [comment]
`"extract-text-webpack-plugin":` "next",` did work for me :D

**Comment by jfsiii** [comment]
@Eddybag That's not a valid `package.json` entry.  

If you use `yarn`, type/paste this on the command line:

`yarn add --dev extract-text-webpack-plugin@next` (see https://github.com/webpack-contrib/extract-text-webpack-plugin/issues/701#issuecomment-392691723)

Otherwise use this for `npm`:

`npm i -D extract-text-webpack-plugin@next` (see https://github.com/webpack-contrib/extract-text-webpack-plugin/issues/701#issuecomment-368475240)

**Comment by satyaprakash-k** [comment]
npm i -D extract-text-webpack-plugin@next worked. Able to bundle now without any errors. Thanks @jfsiii

**Comment by edwinharly** [comment]
I've done this `npm i -D extract-text-webpack-plugin@next` still didn't work out for me.
Here's the log
```
npm run development


> @ development D:\RDPropertyManager\propertymanager
> cross-env NODE_ENV=development node_modules/webpack/bin/webpack.js --progress --hide-modules --config=node_modules/laravel-mix/setup/webpack.config.js

 77% module and chunk tree optimization unnamed compat pluginD:\RDPropertyManager\propertymanager\node_modules\webpack\lib\Chunk.js:750
                throw new Error(
                ^

Error: Chunk.entrypoints: Use Chunks.groupsIterable and filter by instanceof Entrypoint instead
    at Chunk.get (D:\RDPropertyManager\propertymanager\node_modules\webpack\lib\Chunk.js:750:9)
    at D:\RDPropertyManager\propertymanager\node_modules\laravel-mix\node_modules\extract-text-webpack-plugin\dist\index.js:176:48
    at Array.forEach (<anonymous>)
    at D:\RDPropertyManager\propertymanager\node_modules\laravel-mix\node_modules\extract-text-webpack-plugin\dist\index.js:171:18
    at AsyncSeriesHook.eval [as callAsync] (eval at create (D:\RDPropertyManager\propertymanager\node_modules\tapable\lib\HookCodeFactory.js:24:12), <anonymous>:12:1)
    at AsyncSeriesHook.lazyCompileHook [as _callAsync] (D:\RDPropertyManager\propertymanager\node_modules\tapable\lib\Hook.js:35:21)
    at Compilation.seal (D:\RDPropertyManager\propertymanager\node_modules\webpack\lib\Compilation.js:905:27)
    at hooks.make.callAsync.err (D:\RDPropertyManager\propertymanager\node_modules\webpack\lib\Compiler.js:494:17)
    at _err0 (eval at create (D:\RDPropertyManager\propertymanager\node_modules\tapable\lib\HookCodeFactory.js:24:12), <anonymous>:11:1)
    at _addModuleChain (D:\RDPropertyManager\propertymanager\node_modules\webpack\lib\Compilation.js:771:12)
    at processModuleDependencies.err (D:\RDPropertyManager\propertymanager\node_modules\webpack\lib\Compilation.js:710:9)
    at _combinedTickCallback (internal/process/next_tick.js:131:7)
    at process._tickCallback (internal/process/next_tick.js:180:9)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @ development: `cross-env NODE_ENV=development node_modules/webpack/bin/webpack.js --progress --hide-modules --config=node_modules/laravel-mix/setup/webpack.config.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the @ development script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\Edwin\AppData\Roaming\npm-cache\_logs\2018-06-21T05_49_49_534Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @ dev: `npm run development`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the @ dev script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

**Comment by iamzee** [comment]
extract-text-webpack-plugin
version 4.0.0-beta.0 worked for me...

**Comment by mnpenner** [comment]
Still getting infinite compilation with beta.0.

**Comment by PocketNinjaDesign** [comment]
It seems it may be time to move away from extract-text-webpack-plugin

https://github.com/webpack-contrib/mini-css-extract-plugin

haven't used it yet but I'm going to change my project with it!!!

**Comment by terencemutizwa** [comment]
extract-text-webpack-plugin@next worked for me but it now only bundles the last entry point styles if i specify the name _filename:style.bundle.css_ else it gives multiple bundle files if i use _filename:[name].css_

**Comment by jackwilsdon** [comment]
@seniorYsir [This plugin is deprecated since Webpack 4](https://github.com/webpack-contrib/extract-text-webpack-plugin#usage) (and it looks like it's not going to be updated for Webpack 4), you should switch to `mini-css-extract-plugin` instead.

**Comment by drdanmorris** [comment]
If it helps anyone the jump from **extract-text-webpack-plugin** to **mini-css-extract-plugin** was fairly painless. 

`{
				test: /\.css$/,
				use: [
						{
							loader: MiniCssExtractPlugin.loader
						},

						{
							loader: "css-loader",
							options: {
								sourceMap: true,
								modules: true,
								localIdentName: "[local]___[hash:base64:5]"
							}
						},

						{
							loader: 'postcss-loader'
						}
					]
			}`

**Comment by chuckbergeron** [comment]
If you're following the great advice of @drdanmorris don't forget to require the MiniCssExtractPlugin at the top of your config file:

> const MiniCssExtractPlugin = require('mini-css-extract-plugin'); 
> ...
>   plugins: [
>     new MiniCssExtractPlugin()
>   ],
> ```

**Comment by alexkyse1** [comment]
I've migrated webpack 2 to 4, I needed to migrate from **ExtractTextPlugin** to **MiniCssExtractPlugin** I'm struggling to implement multiple css files, I've reached a point where the console doesn't throw errors but my **ReactJS** webpage does not load at all, blank. 

Does **extract-css-chunks-webpack-plugin** have a differnet approach for this concrete case, if so could someone bear a hand to me please?

### WebPack 2 Configuration

```javascript
const webpack = require('webpack');

const resolve = require('path').resolve;

const ExtractTextPlugin = require('extract-text-webpack-plugin');

const HTMLWebpackPlugin = require('html-webpack-plugin');

const CopyWebpackPlugin = require('copy-webpack-plugin');

const extractAppStyles = new ExtractTextPlugin('css/app.[contenthash].css');

const extractVendorStyles = new ExtractTextPlugin('css/vendor.[contenthash].css');

// This contains shared configuration for dev and prod builds
module.exports = {

	entry: {

		// All App source files will be compiled into main
		app: './src/index.js',

		// All vendor files will be compiled into vendor.
		// You should add new packages you install here. 
		vendor: [
			'babel-polyfill',
			'react',
			'react-dom',
			'react-router-dom',
			'semantic-ui-react',
			'dateformat',
			'axios',
			'react-waypoint',
			//'@babel/preset-react'
		]
	},
	devServer: {
		historyApiFallback: true,
		noInfo: true,
		contentBase: './dist',
		host: '0.0.0.0',
		hot: true,
		open: true,
		historyApiFallback: true,
		inline: true
	},
	module: {
		rules: [
			// Transpile all .js and .jsx files
			{
				test: /\.(js|jsx)?$/,
				exclude: /(node_modules)/,
				use: [
					{
						loader: 'babel-loader',
						options: {
							presets: [
								'@babel/react', 
								'@babel/env', 
							],
							plugins: [
								"@babel/plugin-proposal-class-properties", 
								[
									"@babel/plugin-proposal-decorators", { "legacy": true }
								],
								'@babel/plugin-proposal-object-rest-spread', 
								'@babel/plugin-syntax-dynamic-import', 
							],
						}
					},
				],
			},
			// Compile CSS files
			{ test: /\.css$/, loader: "style-loader!css-loader" },
			// Compile SCSS files
			{
				test: /\.scss$/,
				// This compiles styles specific to this app
				include: resolve(__dirname, './src/app/styles'),
				use: extractAppStyles.extract({
					fallback: 'style-loader',
					use: [
						{ loader: 'css-loader', options: { minimize: true, sourceMap: true } },
						{ loader: 'sass-loader', options: { sourceMap: true } },
					]
				}),
			},
			{
				test: /\.scss$/,
				// This compiles styles from Semantic-UI
				include: resolve(__dirname, './src/assets'),
				use: extractVendorStyles.extract({
					fallback: 'style-loader',
					use: [
						{ loader: 'css-loader', options: { minimize: true, sourceMap: true } },
						{ loader: 'sass-loader', options: { sourceMap: true } },
					]
				}),
			},

			// Copy static assets over with file-loader
			{
				test: /\.(ico)(\?v=[0-9]\.[0-9]\.[0-9])?$/, 
				loader: 'file-loader', options: {name: '[name].[ext]'},
			},
			{
				test: /\.(woff|woff2|eot|ttf|otf)(\?v=[0-9]\.[0-9]\.[0-9])?$/, 
				loader: 'file-loader', options: {name: 'fonts/[name].[ext]'},
			},
			{
				test: /\.(jpg|gif|png|svg)(\?v=[0-9]\.[0-9]\.[0-9])?$/, 
				loader: 'file-loader', options: {name: 'images/[name].[ext]'},
			}
		]
	},
	resolve: {
		extensions: ['*', '.js', '.jsx'],
	},
	plugins: [
		// This pulls out webpack module IDs that changes every build to help with caching
		new webpack.HashedModuleIdsPlugin(),
		
		// This separates vendor-provided code into a separate chunk
		new webpack.optimize.CommonsChunkPlugin({
			name: 'vendor'
		}),
		
		// This pulls out webpack boilerplate code that changes every build to help with caching
		new webpack.optimize.CommonsChunkPlugin({
			name: 'runtime'
		}),

		// Extract styles into a separate css files
		extractAppStyles,
		extractVendorStyles,

		// Inject the build date as an environment variable 
		new webpack.DefinePlugin({
			'process.env':{
				'BUILD_DATE': JSON.stringify(new Date())
			}
		}),
		
		// Inject the required assets into the template index file
		new HTMLWebpackPlugin({
			filename: 'index.html',
			template: 'src/assets/index.html',
		}),

		// Copy public files into the dist folder
		new CopyWebpackPlugin([
			{ from: 'src/public' }
		]),
	],
	output: {
		path: __dirname + '/dist',
		filename: 'js/[name].[chunkhash].js',
		chunkFilename: 'js/[name].[chunkhash].js',
		publicPath: '/',
	},
};
```
### WebPack 4 Configuration

```javascript
const webpack = require('webpack');
const resolve = require('path').resolve;
//const ExtractTextPlugin = require('extract-text-webpack-plugin');
const HTMLWebpackPlugin = require('html-webpack-plugin');
const CopyWebpackPlugin = require('copy-webpack-plugin');
const MiniCssExtractPlugin = require("mini-css-extract-plugin");
const devMode = process.env.NODE_ENV !== 'production'

//const extractAppStyles = new ExtractTextPlugin('css/app.[contenthash].css');
//const extractVendorStyles = new ExtractTextPlugin('css/vendor.[contenthash].css');


// This contains shared configuration for dev and prod builds
module.exports = {
	entry: {
		// All App source files will be compiled into main
		app: './src/index.js',

		// All vendor files will be compiled into vendor.
		// You should add new packages you install here. 
		vendor: [
			'babel-polyfill',
			'react',
			'react-dom',
			'react-router-dom',
			'semantic-ui-react',
			'dateformat',
			'axios',
			'react-waypoint',
		]
	},
	devServer: {
		historyApiFallback: true,
		noInfo: true,
		contentBase: './dist',
		host: '0.0.0.0',
		hot: true,
		open: true,
		historyApiFallback: true,
		inline: true
	},
	module: {
		rules: [
			// Transpile all .js and .jsx files
			{
				test: /\.(js|jsx)?$/,
				exclude: /(node_modules)/,
				use: [
					{
						loader: 'babel-loader',
						options: {
							presets: [
								'@babel/react', 
								'@babel/env', 
							],
							plugins: [
								"@babel/plugin-proposal-class-properties", 
								[
									"@babel/plugin-proposal-decorators", { "legacy": true }
								],
								'@babel/plugin-proposal-object-rest-spread', 
								'@babel/plugin-syntax-dynamic-import', 
							],
						}
					},
				],
			},
			// Compile CSS files
			{ test: /\.css$/, use: [MiniCssExtractPlugin.loader, "css-loader"] },
			// Compile SCSS files
			{
				test: /\.scss$/,
				// This compiles styles specific to this app
				include: resolve(__dirname, './src/app/styles'),
				use: [
					devMode ? 'style-loader' : MiniCssExtractPlugin.loader,
					{ loader: 'css-loader', options: { minimize: true, sourceMap: true } },
					{ loader: 'sass-loader', options: { 
						sourceMap: true,
						includePaths: [resolve(__dirname, "css/app.css")]
					 } }
				],
			},
			{
				test: /\.scss$/,
				// This compiles styles from Semantic-UI
				include: resolve(__dirname, './src/assets'),
				use: [
					devMode ? 'style-loader' : MiniCssExtractPlugin.loader,
					{ loader: 'css-loader', options: { minimize: true, sourceMap: true } },
					{ loader: 'sass-loader', options: { 
						sourceMap: true,
						includePaths: [resolve(__dirname, "css/vendor.css")]
					 } }
				]
			},

			// Copy static assets over with file-loader
			{
				test: /\.(ico)(\?v=[0-9]\.[0-9]\.[0-9])?$/, 
				loader: 'file-loader', options: {name: '[name].[ext]'},
			},
			{
				test: /\.(woff|woff2|eot|ttf|otf)(\?v=[0-9]\.[0-9]\.[0-9])?$/, 
				loader: 'file-loader', options: {name: 'fonts/[name].[ext]'},
			},
			{
				test: /\.(jpg|gif|png|svg)(\?v=[0-9]\.[0-9]\.[0-9])?$/, 
				loader: 'file-loader', options: {name: 'images/[name].[ext]'},
			}
		]
	},
	resolve: {
		extensions: ['*', '.js', '.jsx'],
	},
	plugins: [
		// This pulls out webpack module IDs that changes every build to help with caching
		new webpack.HashedModuleIdsPlugin(),
		
		// This separates vendor-provided code into a separate chunk
		// new webpack.optimize.CommonsChunkPlugin({
		// 	name: 'vendor'
		// }),
		
		// // This pulls out webpack boilerplate code that changes every build to help with caching
		// new webpack.optimize.CommonsChunkPlugin({
		// 	name: 'runtime'
		// }),

		// Extract styles into a separate css files
		//extractAppStyles,
		//extractVendorStyles,

		// Inject the build date as an environment variable 
		new webpack.DefinePlugin({
			'process.env':{
				'BUILD_DATE': JSON.stringify(new Date())
			}
		}),
		
		// Inject the required assets into the template index file
		new HTMLWebpackPlugin({
			filename: 'index.html',
			template: 'src/assets/index.html',
		}),

		// Copy public files into the dist folder
		new CopyWebpackPlugin([
			{ from: 'src/public' }
		]),
		new MiniCssExtractPlugin({
			// Options similar to the same options in webpackOptions.output
			// both options are optional
			filename: devMode ? '[name].css' : '[name].[hash].css',
			chunkFilename: devMode ? '[id].css' : '[id].[hash].css',
		}),
		new MiniCssExtractPlugin({
			filename: devMode ? 'app.css' : 'app.[hash].css',
        }),
        new MiniCssExtractPlugin({
            filename: devMode ? 'vendor.css' : 'vendor.[hash].css',
		})
	],
	optimization: {
		splitChunks: {
			chunks: 'all',
		},
		runtimeChunk: true,
	},
	output: {
		path: __dirname + '/dist',
		filename: 'js/[name].[chunkhash].js',
		chunkFilename: 'js/[name].[chunkhash].js',
		publicPath: '/',
	},
};

```
